### PR TITLE
Build different variants for macOS system ABIs (before/after macOS 12.0)

### PR DESCRIPTION
**Pull request by h-vetinari**
With the bump of the default CI image for macOS to 12.0, we're now running into #77 again (see #148). We intentionally want to match the system ABI (for rationale see [here](https://github.com/conda-forge/libcxx-testing-feedstock/tree/main/recipe/pybind11_exception_rtti_test)). Build for both ABIs.

Since we're relying on the system libcxx for the downstream tests, we need to manually override the image selection. Until the deprecated `macOS-11` image is removed ([end of June](https://github.blog/changelog/2024-05-20-actions-upcoming-changes-to-github-hosted-macos-runners/#macos-11-deprecation-and-removal)!), this looks like a reasonable effort to make. After that, I don't know if there's still a way to test against the old system ABI.

Closes #148 
Closes #149

### CODE REVIEW

**Comment by beckermr** [comment]
We need to patch older builds before and after the Abi patch with the same constraints?

**Comment by h-vetinari** [comment]
> We need to patch older builds before and after the Abi patch with the same constraints?

So the good thing is that generally we have no caps anywhere for `libcxx`, meaning that all environments should pull in the fixed builds (after this PR). Given that this issue appeared ~3 years ago (#77), and given further how extremely wide-spread libcxx is, and how little we've heard about this, I'm cautiously tending towards not patching.

If we _do_ repo-patch, we'd IMO have to apply the same split builds to all relevant major version branches, say 14, 15, 16, otherwise there wouldn't be any installable build for those libcxx versions on `__osx>=12`.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -57,6 +60,7 @@ outputs:
       run:
         - {{ pin_subpackage("libcxxabi", exact=True) }}    # [linux]
       run_constrained:
+        - __osx <12                                        # [osx and (sys_abi == "pre-12")]
```
Comment (beckermr) [inline recipe/meta.yaml:63]: Did we want a corresponding >= here too for the newer Abi? I thought we need it both ways.

**Review by beckermr — CHANGES_REQUESTED** [review]

File: recipe/meta.yaml (hunk 1 above)
Comment (h-vetinari) [reply recipe/meta.yaml:63]: The `>=` is provided through `{{ stdlib("c") }}`, which we're setting accodringly in CBC.

Comment (beckermr) [reply recipe/meta.yaml:63]: Ahhhhh yes thank you!

**Review by beckermr — APPROVED** [review]

**Comment by h-vetinari** [comment]
FYI: the macOS-11 images will be deleted end of June. If we care to test #155 for the old ABI before release, we need to merge this PR soon-ish...

**Comment by isuruf** [comment]
Since it's been a few years since macos-12 came up and we didn't see any issues, how about we put `sys_abi < 12` package in `conda-forge/label/libcxx-macos-lt-12` or something like that and remove `__osx` constraint?

After thinking about consequences, I'm just not sure if it's worth introducing a `__osx` virtual package constraint on `libcxx`

**Comment by h-vetinari** [comment]
I don't really mind either way. I'm trying to implement what you had suggested in the other thread.

I'm also a bit surprised how little we've heard of this, but since the ABI break is clearly observable in our tests (which are meant to test relevant scenarios), it's possible that this is either really rare in practice, or people just don't bother to report it.

From the POV of technical/philosophical correctness, I'd prefer to build for both ABIs (and publish to `main`), but then, I don't know what issues you had in mind with:

> After thinking about consequences, I'm just not sure if it's worth introducing a `__osx` virtual package constraint on `libcxx`

**Comment by isuruf** [comment]
> which are meant to test relevant scenarios), 

The test is a little convoluted in that it uses `-Wl,-rpath`, but not `-L`.

> I don't know what issues you had in mind with:

With things like miniforge, we have to be careful to package it with `sys_abi<12` or otherwise we'll be installing a package that has wrong deps in the base environment. Not sure what the consequences of that are.

**Comment by h-vetinari** [comment]
Ok, so takeaway is we believe this problem is basically impossible to hit in normal use, and we are fine to find out the consequences (if any) of breaking the ABI for <12 users by publishing the old ABI to a label only. Are you in agreement with that, or could you correct my understanding if not?

**Comment by isuruf** [comment]
> Ok, so takeaway is we believe this problem is basically impossible to hit in normal use, and we are fine to find out the consequences (if any) of breaking the ABI for <12 users by publishing the old ABI to a label only. Are you in agreement with that

Yes. If there are any consequences, people have a escape hatch with the label.

File: recipe/conda_build_config.yaml [hunk 2]
```diff
@@ -10,14 +10,45 @@ c_compiler_version:     # [linux]
...
+  - "11.0"                  # [osx and arm64]
+  # 12.0 is not available in conda-forge-ci-setup
+  - "12.3"                  # [osx]
+
+c_stdlib_version:           # [linux or osx]
   # actually would need >=2.24, but we don't have that in conda-forge yet
   - "2.17"                  # [linux]
   - "10.13"                 # [osx and x86_64]
+  - "11.0"                  # [osx and arm64]
+  - "12.0"                  # [osx]
```
Comment (isuruf) [inline recipe/conda_build_config.yaml:29]: Let's keep this at `10.13` so that the libcxx in main label can be installed in older macos versions too.

Comment (h-vetinari) [reply recipe/conda_build_config.yaml:29]: Thanks; I'll fix it in #155, resp. when I create the 17.x branch; the `_0` builds of 17.0.6 already are 10.13-compatible.

**Pull request merged**